internal/service/ch/queries.go

```diff
@@ -239,57 +245,79 @@ SELECT
+        WHEN name = 'speed' AND agg = 'MAX' THEN max(value_number)
+        WHEN name = 'obdRunTime' AND agg = 'MEDIAN' THEN median(value_number)
         ELSE NULL
     END AS value_number,
     CASE
-        WHEN name = 'powertrainType' THEN arrayStringConcat(groupUniqArray(value_string),',')
-        WHEN name = 'powertrainFuelSystemSupportedFuelTypes' THEN groupArraySample(1, 1716404995385)(value_string)[1]
+        WHEN name = 'powertrainType' AND agg = 'UNIQUE' THEN arrayStringConcat(groupUniqArray(value_string),',')
+        WHEN name = 'powertrainFuelSystemSupportedFuelTypes' AND agg = 'RAND' THEN groupArraySample(1, 1716404995385)(value_string)[1]
         ELSE NULL
     END AS value_string
 FROM
     `signal`
+JOIN
+	VALUES(
+		'name String, agg String',
+		('speed, 'MAX'),
+		('obdRunTime', 'MEDIAN'),
+		('powertrainType', 'UNIQUE'),
+		('powertrainFuelSystemSupportedFuelTypes', 'RAND')
+	) AS agg_table
+ON
+	signal.name = agg_table.name
 WHERE
-    `name` IN ['speed', 'obdRunTime', 'powertrainType', 'powertrainFuelSystemSupportedFuelTypes']
-    AND token_id = 15
+    token_id = 15
     AND timestamp > toDateTime('2024-04-15 09:21:19')
     AND timestamp < toDateTime('2024-04-27 09:21:19')
 GROUP BY
     group_timestamp,
     name
 ORDER BY
-    group_timestamp ASC;
+    group_timestamp ASC,
+	name ASC,
+	agg ASC;
 */
-func getAggQuery(aggArgs *model.AggregatedSignalArgs) (string, []any) {
+func getAggQuery(aggArgs *model.AggregatedSignalArgs) (string, []any, error) {
 	if aggArgs == nil {
-		return "", nil
+		return "", nil, nil
+	}
+
+	numAggs := len(aggArgs.FloatArgs) + len(aggArgs.StringArgs)
+	if numAggs == 0 {
+		return "", nil, errors.New("no aggregations requested")
 	}
 
-	names := make([]string, 0, len(aggArgs.FloatArgs)+len(aggArgs.StringArgs))
+	valuesArgs := make([]string, 0, numAggs)
 	for _, agg := range aggArgs.FloatArgs {
-		names = append(names, agg.Name)
+		valuesArgs = append(valuesArgs, fmt.Sprintf("('%s', '%s')", agg.Name, agg.Agg))
 	}
 	for _, agg := range aggArgs.StringArgs {
-		names = append(names, agg.Name)
+		valuesArgs = append(valuesArgs, fmt.Sprintf("('%s', '%s')", agg.Name, agg.Agg))
 	}
+	valueTable := fmt.Sprintf("VALUES('%s', %s) as %s ON %s.%s = %s.%s", valueTableDef, strings.Join(valuesArgs, ", "), aggTableName, vss.TableName, vss.NameCol, aggTableName, vss.NameCol)
 
 	mods := []qm.QueryMod{
 		qm.Select(vss.NameCol),
+		qm.Select(AggCol),
 		selectInterval(aggArgs.Interval),
 		selectNumberAggs(aggArgs.FloatArgs),
 		selectStringAggs(aggArgs.StringArgs),
-		qm.WhereIn(nameIn, names),
 		qm.Where(tokenIDWhere, aggArgs.TokenID),
 		qm.Where(timestampFrom, aggArgs.FromTS),
 		qm.Where(timestampTo, aggArgs.ToTS),
 		qm.From(vss.TableName),
+		qm.InnerJoin(valueTable),
 		qm.GroupBy(IntervalGroup),
 		qm.GroupBy(vss.NameCol),
+		qm.GroupBy(AggCol),
 		qm.OrderBy(groupAsc),
+		qm.OrderBy(nameAsc), // These last two order-bys are just to make the order standard for tests.
+		qm.OrderBy(aggAsc),
```

Comment:
Should we sort the results in the test instead of adding more complexity to the query?

Reply:
Let's get rid of it.